### PR TITLE
Custom element upgrades

### PR DESCRIPTION
**Pull request by cbrewster**
---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/17935)

### CODE REVIEW

File: components/script/dom/bindings/interface.rs [hunk 1]
```diff
@@ -281,24 +282,42 @@ pub unsafe fn html_constructor<T>(window: &Window, call_args: &CallArgs) -> Fall
...
+            // Step 12
+            let mut construction_stack = definition.construction_stack.borrow_mut();
+            construction_stack.pop();
+            construction_stack.push(ConstructionStackEntry::AlreadyConstructedMarker);
+
+            // Step 13
+            Root::downcast(element).ok_or(Error::InvalidState)
+        },
+        // Step 10
+        Some(ConstructionStackEntry::AlreadyConstructedMarker) => return Err(Error::InvalidState),
```
Comment (jdm) [inline components/script/dom/bindings/interface.rs]: No need for `return` here.

File: components/script/dom/customelementregistry.rs [hunk 2]
```diff
@@ -462,6 +462,72 @@ impl CustomElementDefinition {
...
+    {
+        // Go into the constructor's compartment
+        let _ac = JSAutoCompartment::new(cx, definition.constructor.callback());
+        let args = HandleValueArray::new();
+        // Step 7
+        if unsafe { !Construct1(cx, constructor.handle(), &args, construct_result.handle_mut()) } {
+            // TODO: Catch exceptions
+            return;
+        }
+        // Step 8
```
Comment (jdm) [inline components/script/dom/customelementregistry.rs]: This is 7.2.

File: components/script/dom/customelementregistry.rs [hunk 3]
```diff
@@ -462,6 +462,72 @@ impl CustomElementDefinition {
...
+    let cx = window.get_cx();
+    rooted!(in(cx) let constructor = ObjectValue(definition.constructor.callback()));
+    rooted!(in(cx) let mut element_val = UndefinedValue());
+    unsafe { element.to_jsval(cx, element_val.handle_mut()); }
+    rooted!(in(cx) let mut construct_result = ptr::null_mut());
+    {
+        // Go into the constructor's compartment
+        let _ac = JSAutoCompartment::new(cx, definition.constructor.callback());
+        let args = HandleValueArray::new();
+        // Step 7
```
Comment (jdm) [inline components/script/dom/customelementregistry.rs]: 7.1

File: components/script/dom/customelementregistry.rs [hunk 4]
```diff
@@ -462,6 +462,72 @@ impl CustomElementDefinition {
...
+        }
+        // Step 8
+        let mut same = false;
+        rooted!(in(cx) let construct_result_val = ObjectValue(construct_result.get()));
+        if unsafe { !JS_SameValue(cx, construct_result_val.handle(), element_val.handle(), &mut same) } {
+            // TODO: Catch exceptions
+            return;
+        }
+        if !same {
+            // TODO: Throw InvalidStateError
```
Comment (jdm) [inline components/script/dom/customelementregistry.rs]: ?

File: components/script/dom/customelementregistry.rs [hunk 5]
```diff
@@ -79,13 +79,19 @@ impl CustomElementRegistry {
 
     /// https://html.spec.whatwg.org/multipage/#look-up-a-custom-element-definition
     pub fn lookup_definition(&self,
-                             local_name: LocalName,
-                             is: Option<LocalName>)
+                             namespace: &Namespace,
+                             local_name: &LocalName,
+                             is: Option<&LocalName>)
                              -> Option<Rc<CustomElementDefinition>> {
+        // Step 1
```
Comment (jdm) [inline components/script/dom/customelementregistry.rs]: Let's move this to Document::lookup_custom_element_definition instead.

File: components/script/dom/customelementregistry.rs [hunk 6]
```diff
@@ -287,17 +288,28 @@ impl CustomElementRegistryMethods for CustomElementRegistry {
...
+
+        // Step 13
+        let document = self.window.Document();
+
+        // Steps 14-15
+        for candidate in document.upcast::<Node>().traverse_preorder().filter_map(Root::downcast::<Element>) {
+            let is = candidate.get_is();
+            if *candidate.local_name() == local_name &&
+                *candidate.namespace() == ns!(html) &&
+                (is.is_none() || is.as_ref() == Some(&name))
```
Comment (jdm) [inline components/script/dom/customelementregistry.rs]: Don't we need to check `extends` here?

File: components/script/dom/customelementregistry.rs [hunk 7]
```diff
@@ -492,48 +492,70 @@ pub fn upgrade_element(definition: Rc<CustomElementDefinition>, element: &Elemen
...
+            report_pending_exception(cx, true);
+        }
+        return;
+    }
+
+    element.set_custom_element_definition(definition);
+}
+
+#[allow(unsafe_code)]
+fn run_upgrade_constructor(constructor: &Rc<Function>, element: &Element) -> ErrorResult {
```
Comment (jdm) [inline components/script/dom/customelementregistry.rs]: Add a spec link here.

File: components/script/dom/element.rs [hunk 8]
```diff
@@ -150,6 +150,7 @@ pub struct Element {
     /// https://dom.spec.whatwg.org/#concept-element-custom-element-definition
     #[ignore_heap_size_of = "Rc"]
     custom_element_definition: DOMRefCell<Option<Rc<CustomElementDefinition>>>,
+    custom_element_state: Cell<CustomElementState>,
```
Comment (jdm) [inline components/script/dom/element.rs]: Add a link to https://dom.spec.whatwg.org/#concept-element-custom-element-state .

File: components/script/dom/element.rs [hunk 9]
```diff
@@ -179,6 +180,14 @@ pub enum CustomElementCreationMode {
     Asynchronous,
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, HeapSizeOf, JSTraceable)]
+pub enum CustomElementState {
```
Comment (jdm) [inline components/script/dom/element.rs]: Add a link to https://dom.spec.whatwg.org/#concept-element-custom-element-state .

**Review by jdm — CHANGES_REQUESTED** [review]
Great work! These changes were quite readable and straightforward to review.

File: components/script/dom/bindings/interface.rs (hunk 1 above)
Comment (cbrewster) [inline components/script/dom/bindings/interface.rs]: Done.

File: components/script/dom/customelementregistry.rs (hunk 4 above)
Comment (cbrewster) [inline components/script/dom/customelementregistry.rs]: Sorry, this was implemented in a later commit.

File: components/script/dom/customelementregistry.rs (hunk 3 above)
Comment (cbrewster) [inline components/script/dom/customelementregistry.rs]: Done.

File: components/script/dom/customelementregistry.rs (hunk 2 above)
Comment (cbrewster) [inline components/script/dom/customelementregistry.rs]: Done.

File: components/script/dom/customelementregistry.rs (hunk 5 above)
Comment (cbrewster) [inline components/script/dom/customelementregistry.rs]: Done.

File: components/script/dom/customelementregistry.rs (hunk 6 above)
Comment (cbrewster) [inline components/script/dom/customelementregistry.rs]: Oops, yup. Fixed.

File: components/script/dom/customelementregistry.rs (hunk 7 above)
Comment (cbrewster) [inline components/script/dom/customelementregistry.rs]: Done.

File: components/script/dom/element.rs (hunk 8 above)
Comment (cbrewster) [inline components/script/dom/element.rs]: Done.

File: components/script/dom/element.rs (hunk 9 above)
Comment (cbrewster) [inline components/script/dom/element.rs]: Done.

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /_mozilla/mozilla/collections.html:
  │ FAIL [expected PASS] test_binding\n    page   43
  │   → assert_true: Should be function HTMLUnknownElement() {
    [native code]
} expected true got false
  │ 
  │ check_collection/<@http://web-platform.test:8000/_mozilla/mozilla/:86:11
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  │ check_collection@http://web-platform.test:8000/_mozilla/mozilla/:71:5
  │ check_tag@http://web-platform.test:8000/_mozilla/mozilla/:95:7
  │ @http://web-platform.test:8000/_mozilla/mozilla/:145:3
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  └ @http://web-platform.test:8000/_mozilla/mozilla/:69:1

  ▶ Unexpected subtest result in /_mozilla/mozilla/prototypes.html:
  │ FAIL [expected PASS] Untitled
  │   → assert_true: Should be HTMLUnknownElement expected true got false
  │ 
  │ @http://web-platform.test:8000/_mozilla/mozilla/prototypes.html:21:3
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1406:20
  │ test@http://web-platform.test:8000/resources/testharness.js:497:9
  └ @http://web-platform.test:8000/_mozilla/mozilla/prototypes.html:10:1
```

**Pull request merged**